lib/src/main/scala/spinal/lib/com/jtag/sim/JtagVpi.scala

```diff
@@ -0,0 +1,158 @@
+            val tdoSeqBytes = tdoSeq.grouped(8).map(_.foldLeft(0)((acc, b) => (acc << 1) | (if (b) 1 else 0))).toArray
+            // copy the result to bufferIn
+            for (i <- 0 until length) {
+              bufferIn(i) = tdoSeqBytes(i).toByte
+            }
+            // create packet to send to the client, exactly the same as the received packet but with the tdoSeq
+            val vpiCmd = VpiCmd(Cmds.SCAN_CHAIN, bufferOut, bufferIn, length, nbBits)
+            serialize(buffer, vpiCmd)
+            outputStream.write(buffer)
+          }
+          case VpiCmd(Cmds.SCAN_CHAIN_FLIP_TMS, bufferOut, bufferIn, length, nbBits) => {
+            // same as SCAN_CHAIN but with the last TMS set to 1
+            val tdiSeq = bufferOut.map(x => (0 until 8).map(i => (x & (1 << i)) != 0)).flatten.take(nbBits)
+            val tdoSeq = driver.doScanChain(tdiSeq, true)
+            val tdoSeqBytes =
+              tdoSeq.grouped(8).map(_.reverse.foldLeft(0)((acc, b) => (acc << 1) | (if (b) 1 else 0))).toArray
+            for (i <- 0 until length) {
+              bufferIn(i) = tdoSeqBytes(i).toByte
+            }
+            val vpiCmd = VpiCmd(Cmds.SCAN_CHAIN_FLIP_TMS, bufferOut, bufferIn, length, nbBits)
+            serialize(buffer, vpiCmd)
+            outputStream.write(buffer)
+          }
+          case VpiCmd(Cmds.STOP_SIMU, _, _, _, _) => {
+            println("Stop simulation")
+            simSuccess()
+          }
+        }
+      }
+    }
+  }
+
+  def deserializeVpiCmd(buffer: Array[Byte]): VpiCmd = {
+    val byteBuffer = ByteBuffer.wrap(buffer)
+    byteBuffer.order(ByteOrder.LITTLE_ENDIAN)
+    val cmd = byteBuffer.getInt match {
+      case 0 => Cmds.RESET
+      case 1 => Cmds.TMS_SEQ
+      case 2 => Cmds.SCAN_CHAIN
+      case 3 => Cmds.SCAN_CHAIN_FLIP_TMS
+      case 4 => Cmds.STOP_SIMU
+    }
+    val bufferOut = new Array[Byte](XFERT_MAX_SIZE)
+    byteBuffer.get(bufferOut)
+    val bufferIn = new Array[Byte](XFERT_MAX_SIZE)
+    byteBuffer.get(bufferIn)
+    val length = byteBuffer.getInt
+    val nbBits = byteBuffer.getInt
+    VpiCmd(cmd, bufferOut, bufferIn, length, nbBits)
+  }
+
+  def XFERT_MAX_SIZE = 512
+
+  def serialize(buffer: Array[Byte], vpiCmd: VpiCmd) = {
+    val byteBuffer = ByteBuffer.wrap(buffer)
+    byteBuffer.order(ByteOrder.LITTLE_ENDIAN)
+    vpiCmd.cmd match {
+      case Cmds.RESET               => byteBuffer.putInt(0)
+      case Cmds.TMS_SEQ             => byteBuffer.putInt(1)
+      case Cmds.SCAN_CHAIN          => byteBuffer.putInt(2)
+      case Cmds.SCAN_CHAIN_FLIP_TMS => byteBuffer.putInt(3)
+      case Cmds.STOP_SIMU           => byteBuffer.putInt(4)
+    }
+    byteBuffer.put(vpiCmd.bufferOut)
+    byteBuffer.put(vpiCmd.bufferIn)
+    byteBuffer.putInt(vpiCmd.length)
+    byteBuffer.putInt(vpiCmd.nbBits)
+  }
+
+  def MaxSizeOfVpiCmd = 4 + 2 * XFERT_MAX_SIZE + 4 + 4
+
+  case class VpiCmd(cmd: Cmds.Cmd, bufferOut: Array[Byte], bufferIn: Array[Byte], length: Int, nbBits: Int)
```

Comment:
why not have cmd objects derive from VpiCmd so that you don't have to ignore many of the parameters further up?

Reply:
I'm unsure what you mean by that?

Reply:
When matching the cmd above you e.g. have
```
case VpiCmd(Cmds.RESET, _, _, _, _) => driver.doResetTap()
```
You could instead have a
```
case class ResetCmd()
case class TmsSeqCmd(bufferOut: Array[Byte], nBits: Int)
```
But that's a bit of I nitpick I guess...

Reply:
I don't think that would be more readable, and I don't think it would perform better either, so I'd rather not switch it just so it can be more idiomatic